### ISSUE TITLE
About IoU_Smooth L1 loss

### ISSUE THREAD

**Issue by a5372935** [DetectionTeamUCAS/R2CNN-Plus-Plus_Tensorflow#65]
I want to add IoU Smooth L1 loss to SCRDet

```
def iou_smooth_l1_loss_rcnn(bbox_pred, bbox_targets, label, num_classes, sigma=1.0):
    '''

    :param bbox_pred: [-1, (cfgs.CLS_NUM +1) * 5]
    :param bbox_targets:[-1, (cfgs.CLS_NUM +1) * 5]
    :param label:[-1]
    :param num_classes:
    :param sigma:
    :return:
    '''

    outside_mask = tf.stop_gradient(tf.to_float(tf.greater(label, 0)))
    
    # indices = tf.reshape(tf.where(tf.equal(label, 1)), [-1, ])
    # bbox_pred = tf.gather(bbox_pred, indices)

    bbox_pred = tf.reshape(bbox_pred, [-1, num_classes, 5])
    bbox_targets = tf.reshape(bbox_targets, [-1, num_classes, 5])

    value = _smooth_l1_loss_base(bbox_pred,
                                 bbox_targets,
                                 sigma=sigma)
    value = tf.reduce_sum(value, 2)
    value = tf.reshape(value, [-1, num_classes])

    overlaps = tf.py_func(iou_rotate.iou_rotate_calculate2,
                          inp=[tf.reshape(bbox_pred, [-1, 5]), tf.reshape(bbox_targets[:, :-1], [-1, 5])],
                          Tout=[tf.float32])

    overlaps = tf.reshape(overlaps, [-1, 1])
    iou_factor = tf.stop_gradient(-1 * tf.log(overlaps)) / (tf.stop_gradient(value) + cfgs.EPSILON)

    inside_mask = tf.one_hot(tf.reshape(label, [-1, 1]),
                             depth=num_classes, axis=1)

    inside_mask = tf.stop_gradient(
        tf.to_float(tf.reshape(inside_mask, [-1, num_classes])))

    normalizer = tf.to_float(tf.shape(bbox_pred)[0])
    bbox_loss = tf.reduce_sum(
        tf.reduce_sum(value * inside_mask * iou_factor, 1) * outside_mask) / normalizer

    return bbox_loss
```

Is this correct?

**Comment by a5372935** [comment]
> I want to add IoU Smooth L1 loss to SCRDet
> 
> ```
> def iou_smooth_l1_loss_rcnn(bbox_pred, bbox_targets, label, num_classes, sigma=1.0):
>     '''
> 
>     :param bbox_pred: [-1, (cfgs.CLS_NUM +1) * 5]
>     :param bbox_targets:[-1, (cfgs.CLS_NUM +1) * 5]
>     :param label:[-1]
>     :param num_classes:
>     :param sigma:
>     :return:
>     '''
> 
>     outside_mask = tf.stop_gradient(tf.to_float(tf.greater(label, 0)))
>     
>     # indices = tf.reshape(tf.where(tf.equal(label, 1)), [-1, ])
>     # bbox_pred = tf.gather(bbox_pred, indices)
> 
>     bbox_pred = tf.reshape(bbox_pred, [-1, num_classes, 5])
>     bbox_targets = tf.reshape(bbox_targets, [-1, num_classes, 5])
> 
>     value = _smooth_l1_loss_base(bbox_pred,
>                                  bbox_targets,
>                                  sigma=sigma)
>     value = tf.reduce_sum(value, 2)
>     value = tf.reshape(value, [-1, num_classes])
> 
>     overlaps = tf.py_func(iou_rotate.iou_rotate_calculate2,
>                           inp=[tf.reshape(bbox_pred, [-1, 5]), tf.reshape(bbox_targets[:, :-1], [-1, 5])],
>                           Tout=[tf.float32])
> 
>     overlaps = tf.reshape(overlaps, [-1, 1])
>     iou_factor = tf.stop_gradient(-1 * tf.log(overlaps)) / (tf.stop_gradient(value) + cfgs.EPSILON)
> 
>     inside_mask = tf.one_hot(tf.reshape(label, [-1, 1]),
>                              depth=num_classes, axis=1)
> 
>     inside_mask = tf.stop_gradient(
>         tf.to_float(tf.reshape(inside_mask, [-1, num_classes])))
> 
>     normalizer = tf.to_float(tf.shape(bbox_pred)[0])
>     bbox_loss = tf.reduce_sum(
>         tf.reduce_sum(value * inside_mask * iou_factor, 1) * outside_mask) / normalizer
> 
>     return bbox_loss
> ```
> 
> Is this correct?

@yangxue0827  請問 我更改成這樣後是不是比較正確，這樣能順利train但 我現在訓練到15萬次還無法達到正確的旋轉矩形

```
def iou_smooth_l1_loss_rcnn(bbox_pred, bbox_targets, bbox_anchor, anchor_bbox, targets, label, num_classes, sigma=1.0):
    '''

    :param bbox_pred: [-1, (cfgs.CLS_NUM +1) * 5]     pred_offset
    :param bbox_targets:[-1, (cfgs.CLS_NUM +1) * 5]   target_offet
    :param targets:[-1, 6] gt_bbox
    :param label:[-1]
    :param num_classes:
    :param sigma:
    :return:
    '''

    x_c = (anchor_bbox[:, 2] + anchor_bbox[:, 0]) / 2
    y_c = (anchor_bbox[:, 3] + anchor_bbox[:, 1]) / 2
    h = anchor_bbox[:, 2] - anchor_bbox[:, 0] + 1
    w = anchor_bbox[:, 3] - anchor_bbox[:, 1] + 1
    theta = -90 * tf.ones_like(x_c)
    anchor_bbox = tf.transpose(tf.stack([x_c, y_c, w, h, theta]))

    outside_mask = tf.stop_gradient(tf.to_float(tf.greater(label, 0)))

    boxes_pred = bbox_transform.rbbox_transform_inv(boxes=anchor_bbox, deltas=bbox_pred, scale_factors=cfgs.ROI_SCALE_FACTORS)
    boxes_targets = bbox_transform.rbbox_transform_inv(boxes=anchor_bbox, deltas=bbox_targets, scale_factors=cfgs.ROI_SCALE_FACTORS)

    bbox_pred = tf.reshape(bbox_pred, [-1, num_classes, 5])
    bbox_targets = tf.reshape(bbox_targets, [-1, num_classes, 5])

    

    value = _smooth_l1_loss_base(bbox_pred,
                                 bbox_targets,
                                 sigma=sigma)
    value = tf.reduce_sum(value, 2)
    value = tf.reshape(value, [-1, num_classes])

    overlaps = tf.py_func(iou_rotate.iou_rotate_calculate2,
                        inp = [boxes_pred, boxes_targets],
                        Tout=[tf.float32])

    overlaps = tf.reshape(overlaps, [-1, 1])
    iou_factor = tf.stop_gradient(-1 * tf.log(overlaps)) / (tf.stop_gradient(value) + cfgs.EPSILON)

    inside_mask = tf.one_hot(tf.reshape(label, [-1, 1]),
                             depth=num_classes, axis=1)

    inside_mask = tf.stop_gradient(
        tf.to_float(tf.reshape(inside_mask, [-1, num_classes])))

    normalizer = tf.to_float(tf.shape(bbox_pred)[0])
    bbox_loss = tf.reduce_sum(
        tf.reduce_sum(value * inside_mask * iou_factor, 1) * outside_mask) / normalizer

    return bbox_loss
```
